src/wayland/dmabuf/dispatch.rs

```diff
@@ -97,22 +182,29 @@ where
         let data = DmabufData {
             formats: global_data.formats.clone(),
             id: global_data.id,
+            default_feedback: global_data.default_feedback.clone(),
+            known_default_feedbacks: global_data.known_default_feedbacks.clone(),
         };
 
         let zwp_dmabuf = data_init.init(resource, data);
 
         // Immediately send format info to the client if we are the correct version.
         //
         // These events are deprecated in version 4 of the protocol.
-        if zwp_dmabuf.version() <= 3 {
-            for format in &*global_data.formats {
-                zwp_dmabuf.format(format.code as u32);
-
-                if zwp_dmabuf.version() == 3 {
-                    let modifier_hi = (Into::<u64>::into(format.modifier) >> 32) as u32;
-                    let modifier_lo = Into::<u64>::into(format.modifier) as u32;
+        if zwp_dmabuf.version() < zwp_linux_dmabuf_v1::REQ_GET_DEFAULT_FEEDBACK_SINCE {
+            for (fourcc, modifiers) in &*global_data.formats {
+                // Modifier support got added in version 3
+                if zwp_dmabuf.version() < zwp_linux_dmabuf_v1::EVT_MODIFIER_SINCE
+                    && modifiers.contains(&Modifier::Invalid)
+                {
+                    zwp_dmabuf.format(*fourcc as u32);
```

Comment:
I don't think you are supposed to never send `format` on v3, the event was only deprecated on v4.

Reply:
That may be a bit under-defined in the spec, but neither `wlroots` nor `weston` send both. Both either send the format if version is less than modifier since or the modifiers otherwise. Though it seems weston sends the format if it is invalid or linear while wlroots only send the format if it includes invalid.

Reply:
Ok sure, then lets keep it like this.